### DISCUSSION TITLE
Why PassNode::write creates a new resource?

### DISCUSSION THREAD

**Discussion by BrodaJarek** [google/filament#3406] [General]
I am trying to understand why `PassNode::Write()` creates a new resource. I've read EA article and they say `This allows us to catch errors when resources are modified in undefined order (when the same resource is written by different passes). Renaming resources enforce a specific execution order of the render passes.`. The passes are executed in a loop, and I don't see any threading between them, so how can they be executed in undefined order?

**Comment by romainguy** [comment]
@pixelflinger was just explaining this to someone else yesterday :)

"It returns a new handle to the resource

This is to prevent you from making circles"
